consensus/bor/bor.go

```diff
@@ -1170,44 +1156,50 @@ func (c *Bor) GetPendingStateProposals(snapshotNumber uint64) ([]*big.Int, error
 func (c *Bor) CommitStates(
 	state *state.StateDB,
 	header *types.Header,
-	chain core.ChainContext,
+	chCtx chainContext,
 ) error {
-	// get pending state proposals
-	stateIds, err := c.GetPendingStateProposals(header.Number.Uint64() - 1)
-	if err != nil {
-		return err
-	}
-
-	// state ids
-	if len(stateIds) > 0 {
-		log.Debug("Found new proposed states", "numberOfStates", len(stateIds))
-	}
-
-	method := "commitState"
-
-	// itereate through state ids
-	for _, stateID := range stateIds {
-		// fetch from heimdall
-		response, err := c.HeimdallClient.FetchWithRetry("clerk", "event-record", strconv.FormatUint(stateID.Uint64(), 10))
+	number := header.Number.Uint64()
+	if number < c.config.Sprint {
+		return errors.New("Requested to commit states too soon")
+	}
+
+	from := c.EventsLastFetchededAt
+	to := time.Now().Unix()
+	log.Info(
+		"Fetching state updates from Heimdall",
+		"from", time.Unix(from, 0).Format(time.RFC3339),
+		"to", time.Unix(to, 0).Format(time.RFC3339))
+
+	page := 1
+	var eventRecords []EventRecord
+	for true {
+		page++
+		query := fmt.Sprintf("clerk/event-record/list?from-time=%d&to-time=%d&page=%d&limit=%d", from, to, page, stateFetchLimit)
+		log.Info("State events fetch query", "query", query)
+		response, err := c.HeimdallClient.FetchWithRetry(query)
 		if err != nil {
 			return err
 		}
-
-		// get event record
-		var eventRecord EventRecord
-		if err := json.Unmarshal(response.Result, &eventRecord); err != nil {
+		var _eventRecords []EventRecord
+		if err := json.Unmarshal(response.Result, &eventRecords); err != nil {
 			return err
 		}
-
-		// check if chain id matches with event record
-		if eventRecord.ChainID != c.chainConfig.ChainID.String() {
-			return fmt.Errorf(
-				"Chain id proposed state in span, %s, and bor chain id, %s, doesn't match",
-				eventRecord.ChainID,
-				c.chainConfig.ChainID,
-			)
+		eventRecords = append(eventRecords, _eventRecords...)
+		if len(_eventRecords) < stateFetchLimit {
+			break
 		}
+	}
+	sort.SliceStable(eventRecords, func(i, j int) bool {
+		return eventRecords[i].ID < eventRecords[i].ID
+	})
 
+	for _, eventRecord := range eventRecords {
+		// validateEventRecord checks whether an event lies in the specified time range
+		// since the events are sorted by id and if it turns out that event i lies outside the time range,
+		// there is a possibility that all subsequent events lie outside of the time range. Hence we return the err and don't probe any further
+		if err := validateEventRecord(eventRecord, number, from, to, c.chainConfig.ChainID.String()); err != nil {
+			return err
+		}
```

Comment:
Shouldn't we just ignore instead of `err`?

Reply:
got it. breaking loop instead